hbase-client/src/main/java/org/apache/hadoop/hbase/client/AsyncNonMetaRegionLocator.java

```diff
@@ -577,6 +635,15 @@ private void removeLocationFromCache(HRegionLocation loc) {
       if (!canUpdateOnError(loc, oldLoc)) {
         return;
       }
+      // Tell metaReplicaSelector that the location is stale. It will create a stale entry
+      // with timestamp internally. Next time the client looks up the same location,
+      // it will pick a different meta replica region. For the current implementation,
+      // the metaReplicaId is not used, so the primary one is passed in.
+      if (this.metaReplicaMode == CatalogReplicaMode.LoadBalance) {
+        // metaReplicaId is not used in simpleSelector, default to the primary one.
+        metaReplicaSelector.onError(loc, RegionInfo.DEFAULT_REPLICA_ID);
```

Comment:
This is a bit strange? We are using a CatalogReplicaLoadBalanceSelector, not a CatalogReplicaLoadBalanceSimpleSelector, so I do not think we should assume it is a simple selector and pass default replica id here? It may spend a bunch of time for later developers who implement new selectors to find out why the selector does not work as expected.  If you want to do this please just declare a CatalogReplicaLoadBalanceSimpleSelector above.

Reply:
In my initial patch, I do keep "fromReplicaId" in TableCache to indicate which meta replica id the location comes from.
However, this is not being used at this moment so I do not think it is a good idea to add more bytes which is not being used at this moment. How about passing an invalid replica id (say -1) to indicate it is not being filled at this moment?

Reply:
What do you think to bring my initial implementation (fromReplicaId) back so fromReplicaId is going to be a valid one (not faked one)?

Reply:
I pushed a patch with fromReplicaId set to -1. Work on an add-on patch which will fill in this metaReplicaId.

Reply:
As discussed, removed fromReplicaId for now. Will go back to this part later when it is needed.